pkg/azuredisk/azuredisk_v2.go

```diff
@@ -297,6 +298,7 @@ func (d *DriverV2) StartControllersAndDieOnExit(ctx context.Context) {
 	}
 
 	sharedState := controller.NewSharedState()
+	sharedState.RegisterCachedClient(mgr.GetClient())
```

Comment:
Consider passing the `mgr.GetClient()` to the `NewSharedState()` function instead of through a new interface method.

Reply:
I actually meant to insert this line once the controller manager acquires the leader lock so that the cached client is not registered for those that failed to acquire the lock. 

https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/6578e37c53386b1f31358e180d1f207bf371bae6/pkg/azuredisk/azuredisk_v2.go#L356

Wouldn't it make sense to initialize sharedState first and post-append cached client in that case?

Reply:
Nvm. As long as we register sharedState to crdProvisioner only when leader lock is acquired, that should do.

Reply:
Ideally, `sharedState` should be fully initialized (including setting the cached client) before being passed to its dependents through their constructor functions. Post-initialization inject of objects in makes it difficult to track down dependencies and make nil-reference bugs possible since it may not be clear at the point of use whether one can assume it was initialized. Consumers of this field have to deal with the possibility that it may be nil, increasing complexity.

I am equally concerned about the `RegisterSharedState` method on the `CrdProvisioner` but I don't see a good way of handling this unless we drastically change how the driver and controllers are initialized.